Add option for users to change their own role.

whilstsomebody opened the pull request:
Previously, users who were permitted to change their own role had no way to do so, as there was no setting available for this functionality.

This PR does the following tasks:

- Adds a select dropdown of the various roles available.
- The dropdown in any state shows the current role of the user by default.
- If the user has permission to change their role, the dropdown is  enabled , else disabled.
   - If the user is an owner, then they can change their role only if the org has more than 1 owners.
   - Admins can change their role anytime.
   - For admins, "owner" option is not available in the dropdown.
   - Other users do't have this permission.
- label : role, is followed by a "?" that links to `help/user-roles`.

**Note:** We are sending a `PATCH` request to `/users/<user-id>` with data consisting of new role value obtained from the dropdown.

Fixes: #34829

## Screenshots

- **Owner view: When there is only owner**

![image](https://github.com/user-attachments/assets/36d368dd-2806-44c2-b925-6aa338facb46)

- **Owner view: When there are > 1 owners**

![image](https://github.com/user-attachments/assets/5d682b5e-4142-4ddc-a6ed-78243a4ea920)

- **Admin view: Anytime**

![image](https://github.com/user-attachments/assets/ebff52c1-159c-4572-97bc-aee5edafd9a0)

- **For any other user**

![image](https://github.com/user-attachments/assets/d51d63ec-2b9c-4c0c-abe2-542b20a45750)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

apoorvapendse reviewed (commented):
I got into this weird state when I changed permissions to a role that cannot change roles further (from Iago as admin to a mod).


https://github.com/user-attachments/assets/3da9505d-fa1a-4e52-bcce-efe0af7781ec


We should probably lock/disable the dropdown if someone changes to a role that cannot further modify roles, by possibly re-rendering the overlay (not sure).

Also should we have some sort of confirmation for this? It might lock someone out until they contact the owner due to some misclick or other mishaps.
cc @alya for thoughts.

apoorvapendse reviewed (commented):
I wonder if the list options should also be updated when you switch from owner to admin
I still see the option to become owner on turning to admin and getting that error.
![image](https://github.com/user-attachments/assets/1cfa76d9-053c-4e9d-8f91-30d70f7220f6)

apoorvapendse reviewed (approved):
Looks good.

apoorvapendse commented:
@zulipbot remove "buddy review"

afeefuddin commented on web/src/settings.ts:
```diff
@@ -40,6 +40,32 @@ function user_can_change_password(): boolean {
     return realm.realm_email_auth_enabled;
 }
 
+function user_can_change_role(): boolean {
+    const user = people.get_user_by_id_assert_valid(current_user.user_id);
+    if (user.is_owner && !people.is_current_user_only_owner()) {
+        return true;
+    } else if (user.is_admin && !user.is_owner) {
+        return true;
+    }
+    return false;
+}
```
Why are you getting the user again? Can't we do: 
```suggestion

function user_can_change_role(): boolean {
    if (current_user.is_owner && !people.is_current_user_only_owner()) {
        return true;
    } else if (current_user.is_admin && !current_user.is_owner) {
        return true;
    }
    return false;
}
```

afeefuddin reviewed (approved):
lgtm

alya commented:
Let's have @sahil839 take this one.

sahil839 commented on web/src/settings.ts:
```diff
@@ -40,6 +40,31 @@ function user_can_change_password(): boolean {
+    } else if (current_user.is_admin && !current_user.is_owner) {
+        return true;
+    }
+    return false;
```
I think this can be simplified as -
```js
function user_can_change_role(): boolean {
    if (people.is_current_user_only_owner()) {
        return false;
    }
    return current_user.is_admin;
}
```

sahil839 replied:
Also, this function should be moved to `settings_data` where we have other similar functions and we can probably name this function as `user_can_change_their_own_role`.

sahil839 commented on web/src/settings.ts:
```diff
@@ -155,6 +182,12 @@ export function build_page(): void {
+    const current_user_role = people.get_by_user_id(current_user.user_id).role;
+    $("#user-self-role-select").val(current_user_role);
+    if (!current_user.is_owner) {
+        remove_owner_from_role_select();
```
This code should be in `settings_account.set_up` where we initialize other elements on the page.

sahil839 commented on web/src/settings.ts:
```diff
@@ -40,6 +40,31 @@ function user_can_change_password(): boolean {
+            `option[value="${CSS.escape(settings_config.user_role_values.owner.code.toString())}"]`,
+        )
+        .hide();
+}
```
I guess both the functions above can be moved to `settings_account.ts`.

sahil839 commented on web/src/user_events.ts:
```diff
@@ -119,6 +119,7 @@ export const update_person = function update(event: UserUpdate): void {
             settings_org.maybe_disable_widgets();
             settings_org.enable_or_disable_group_permission_settings();
             settings.update_lock_icon_in_sidebar();
+            settings_account.remove_owner_from_role_select();
```
I think it would be good to add code to live-update the disabled dropdown especially when a current owner makes another user owner so that they can change their role themselves.

sahil839 replied:
This will remove owner option from dropdown even if user is made an owner.

sahil839 replied:
This hasn't been fixed.
> This will remove owner option from dropdown even if user is made an owner.

And also the dropdown is not live-updated to show the updated role.

sahil839 commented on web/src/settings.ts:
```diff
@@ -40,6 +40,21 @@ function user_can_change_password(): boolean {
     return realm.realm_email_auth_enabled;
 }
 
+function user_can_change_their_own_role(): boolean {
```
I think we can move this to `people.ts` then to avoid import cycles.

sahil839 commented on web/templates/settings/account_settings.hbs:
```diff
@@ -69,6 +69,22 @@
+                      id="user-self-role-select"
+                      data-setting-widget-type="number"
+                      {{#unless user_can_change_role}}disabled="disabled"{{/unless}}>
+                        {{> dropdown_options_widget option_values=user_role_values}}
```
I am not sure but tooltip might be helpfule when the dropdown is disabled. Can be discussed first.

whilstsomebody replied:
I was also thinking about this.
Asked in the thread: [#user questions > ownership transfer issue @ 💬](https://chat.zulip.org/#narrow/channel/138-user-questions/topic/ownership.20transfer.20issue/near/2199169)

whilstsomebody commented:
Hey @sahil839 , I have made all the suggested changes.

- Implemented live update for the case you pointed out here: https://github.com/zulip/zulip/pull/34861#discussion_r2155999593

Affected area: https://github.com/zulip/zulip/pull/34861/files#diff-29ec85d51ba1450fe3a26844bba170a480a215a9e5a07aebf1f8439bee47cd81R138-R139

Also, for this to work properly, I have to make some changes in the function `user_can_change_their_own_role`.
In your code, we were returning only `current_user.is_admin`, but after the change, now we return `current_user.is_owner` too.

- Yes (https://github.com/zulip/zulip/pull/34861#discussion_r2155999965) this worked and I also moved other function as suggested.  Moved `user_can_change_their_own_role()` to `people.ts`; `update_user_self_role_select_state()` and `remove_owner_from_role_select()` to `settings_account.ts`.

- https://github.com/zulip/zulip/pull/34861#discussion_r2156000533: Added tooltip for the case when the select dropdown is disabled for the owner only as discussed here: [#user questions > ownership transfer issue @ 💬](https://chat.zulip.org/#narrow/channel/138-user-questions/topic/ownership.20transfer.20issue/near/2199781)

Affected area: https://github.com/zulip/zulip/pull/34861/files#diff-3457b66433da59804baf3995b17aedb32e705e9b98129536cbf421fd9168130fR276-R292

- Added test for testing `user_can_change_their_own_role` in `people.test.cjs`.

Please take a look.

sahil839 commented on web/src/user_events.ts:
```diff
@@ -131,6 +132,11 @@ export const update_person = function update(event: UserUpdate): void {
+            settings_account.update_user_self_role_select_state();
+        }
+
+        if (!people.is_my_user_id(event.user_id)) {
```
I guess this condition can be modified to add check for if `is_owner` value was changed or not.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -272,6 +273,32 @@ export function hide_confirm_email_banner(): void {
+        if (people.is_current_user_only_owner() && current_user.is_owner) {
+            ui_util.disable_element_and_add_tooltip(
+                $select_elem,
+                "The owner role cannot be removed from the only organization owner.",
```
Tooltip is not shown initially because this function is only called from the live-update code.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -272,6 +273,32 @@ export function hide_confirm_email_banner(): void {
+    const $select_elem = $("select#user-self-role-select");
+    if ($select_elem.has("data-tippy-content")) {
+        ui_util.enable_element_and_remove_tooltip($select_elem);
+    }
```
I think this function can be simplified -
```js
if (people.user_can_change_their_own_role()) {
    ui_util.enable_element_and_remove_tooltip();
    return;
}

if (current_user.is_owner) {
    ui_util.disable_element_and_add_tooltip(
        $select_elem,
        "The owner role cannot be removed from the only organization owner.",
    )
} else {
    $select_elem.disabled("true"):
}
```

sahil839 commented on web/src/user_events.ts:
```diff
@@ -131,6 +132,11 @@ export const update_person = function update(event: UserUpdate): void {
             settings_realm_user_settings_defaults.maybe_disable_widgets();
             settings_account.update_account_settings_display();
             settings.update_lock_icon_in_sidebar();
+            settings_account.update_user_self_role_select_state();
```
We should also add the code for actually updating the role dropdown when role of a user is changed.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -909,4 +938,26 @@ export function set_up(): void {
+
+    // Set the default role of the user.
+    const current_user_role = people.get_by_user_id(current_user.user_id).role;
+    $("#user-self-role-select").val(current_user_role);
```
This can be removed now since `update_user_self_role_select_state` handles that.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -272,6 +273,34 @@ export function hide_confirm_email_banner(): void {
     $("#account-settings-status").hide();
 }
 
+export function update_user_self_role_select_state(): void {
```
Would be good to name this as `update_user_own_role_select_state`.

sahil839 reviewed (commented):
The `is_owner` check is needed because on changing other's user role, a user's select element will be needed only if they are owner and the other user either becomes a owner or is removed from being a owner. You do not need to check the `current_user` value you need to check the previous and current value of the user whose role is being changed.

sahil839 commented on web/src/user_events.ts:
```diff
@@ -131,13 +133,14 @@ export const update_person = function update(event: UserUpdate): void {
-            current_user.is_moderator = user.is_moderator;
+        if (!people.is_my_user_id(event.user_id)) {
+            const is_owner = event.role === settings_config.user_role_values.owner.code;
+            if (was_owner || is_owner) {
```
Should this be `was_owner !== user.is_owner`?

sahil839 commented on web/templates/settings/account_settings.hbs:
```diff
@@ -68,6 +68,22 @@
+                      class="prop-element settings_select bootstrap-focus-style"
+                      id="user-self-role-select"
+                      data-setting-widget-type="number"
+                      {{#unless user_can_change_role}}disabled="disabled"{{/unless}}>
```
This can be removed now since disabling is handled in JS.

whilstsomebody commented:
Hey @sahil839 , I have fixed the live update and other issues.

[Screencast from 2025-06-25 14-51-33.webm](https://github.com/user-attachments/assets/4f6057e6-e43f-429d-82b3-a4c9c60525d1)

Please take a look.

whilstsomebody commented:
I have updated the tooltip content as suggested by @alya here : [#user questions > ownership transfer issue @ 💬](https://chat.zulip.org/#narrow/channel/138-user-questions/topic/ownership.20transfer.20issue/near/2204611)

sahil839 commented on web/src/settings.ts:
```diff
@@ -143,6 +143,8 @@ export function build_page(): void {
         email_address_visibility_values: settings_config.email_address_visibility_values,
         owner_is_only_user_in_organization: people.get_active_human_count() === 1,
         user_can_change_password: user_can_change_password(),
+        user_can_change_role: people.user_can_change_their_own_role(),
```
This can be removed now.

sahil839 commented on web/src/user_events.ts:
```diff
@@ -131,13 +133,12 @@ export const update_person = function update(event: UserUpdate): void {
             settings_realm_user_settings_defaults.maybe_disable_widgets();
             settings_account.update_account_settings_display();
             settings.update_lock_icon_in_sidebar();
+            settings_account.update_user_own_role_select_state();
```
And the code to disable the dropdown and update the tooltip on it should be called -
- If current user's role is changed and either is_admin value is changed or is_owner value is changed.
- If other user's role is changed, `is_owner` value is changed.

You can decide these conditions by seeing what values are being used in the respective function, i.e. settings_account.update_user_own_role_select_state and then it should be called if any values used in that function are changed.

whilstsomebody commented on web/src/user_events.ts:
```diff
@@ -131,13 +133,12 @@ export const update_person = function update(event: UserUpdate): void {
-            people.is_my_user_id(event.user_id) &&
-            current_user.is_moderator !== user.is_moderator
-        ) {
-            current_user.is_moderator = user.is_moderator;
```
Oops! Mistake.

whilstsomebody commented:
Hey @sahil839 , I have fixed the code and investigated all the cases of live-update. 
Please take a look.

whilstsomebody commented:
Hey @sahil839 , I have done a lot of print debugging and played around various cases and made some changes in the live-update code. Please see if the new organization is correct or not.
Thanks!

sahil839 commented on web/src/user_events.ts:
```diff
@@ -113,12 +114,16 @@ export const update_person = function update(event: UserUpdate): void {
         user_profile.update_profile_modal_ui(user, event);
+        settings_account.set_user_own_role_dropdown_value();
+        settings_account.add_or_remove_owner_from_role_select();
+        settings_account.update_user_own_role_select_state();
```
I think only the function to set dropdown value should be called for all cases. If not can you specify the scenario? As per my thinking, others are only needed for certain cases like removing "Owner" option is only needed if `is_owner` is changed and updating dropdown disabled state and tooltip is needed as explained [here](https://github.com/zulip/zulip/pull/34861/files#r2168870364).

whilstsomebody commented:
@sahil839 , I have made the changes. 

- [Here](https://github.com/zulip/zulip/pull/34861/files#diff-29ec85d51ba1450fe3a26844bba170a480a215a9e5a07aebf1f8439bee47cd81), I removed `add_or_remove_owner_from_role_dropdown` and `update_user_own_role_dropdown_state` and have kept only `set_user_own_role_dropdown_value`.
- [Here](https://github.com/zulip/zulip/pull/34861/files#diff-29ec85d51ba1450fe3a26844bba170a480a215a9e5a07aebf1f8439bee47cd81R123-R125), added `add_or_remove_owner_from_role_dropdown` because of the error that occurs when role is changed from `owner` to `guest`.
- [Here](https://github.com/zulip/zulip/pull/34861/files#diff-29ec85d51ba1450fe3a26844bba170a480a215a9e5a07aebf1f8439bee47cd81R138-R139) I removed `add_or_remove_owner_from_role_dropdown` because of the error that occurs when role is changed from `admin` to `guest`.
- [Here](https://github.com/zulip/zulip/pull/34861/files#diff-29ec85d51ba1450fe3a26844bba170a480a215a9e5a07aebf1f8439bee47cd81R141), negated the first check. 


I had thought of doing https://github.com/zulip/zulip/pull/34861#discussion_r2174589507 eariler but when I tried, I got certain errors when switching rols between `owner and guest` and `admin and guest` which I was able to fix by doing https://github.com/zulip/zulip/pull/34861#discussion_r2174591470. But I didn't try doing https://github.com/zulip/zulip/pull/34861#discussion_r2174591470 because of https://github.com/zulip/zulip/pull/34861#discussion_r2168861567. 


Please take a look. 
Thanks!

sahil839 commented on web/src/user_events.ts:
```diff
@@ -131,6 +135,11 @@ export const update_person = function update(event: UserUpdate): void {
+            settings_account.update_user_own_role_dropdown_state();
+        }
+
+        if (!people.is_my_user_id(event.user_id) && was_owner !== user.is_owner) {
```
This should also have a condition `current_user.is_owner` since the dropdown update is required when other user's role is changed only for owners.

sahil839 reviewed (commented):
Posted a couple of comments, which I guess should be the final batch of comments from my side.

whilstsomebody commented:
@sahil839 , I have addressed the above two comments. Please take a look.

alya commented:
Can you add a screenshot of the disabled tooltip to the PR description?

> Because you are the only organization owner, you cannot change your role.

alya commented:
Can you take a pass at adding help center documentation to this PR?

It can be a "Change your role" section just below https://zulip.com/help/user-roles#change-a-users-role, with a "Desktop/Web" instructions tab; no text before the instructions. You can take a look at, e.g., instructions for changing your email for reference.

alya commented on help/user-roles.md:
```diff
@@ -75,6 +75,25 @@ existing owner's role.
+{settings_tab|account-and-privacy}
+
+1. Under **Account**, select a [role](#roles) from the **Roles** dropdown.
+   If the dropdown is disabled in your case, you do not have permission to
+   to change you role.
```
```suggestion
   If the dropdown is disabled in your case, you do not have permission to
   to change you role.
```

Not needed, as we already have an admin-only banner.

whilstsomebody commented:
@alya , how's this?

![image](https://github.com/user-attachments/assets/2c135b1b-752a-479a-85c9-e213053622eb)

alya commented:
@timabbott This PR was reviewed by @sahil839 ; should be ready for you.

sahil839 commented:
Screenshot for the new help center change -

<img width="821" height="350" alt="image" src="https://github.com/user-attachments/assets/1dea4d36-19ef-4bdd-914a-94095a09f851" />
